### PR TITLE
Add location for the calling account

### PR DESCRIPTION
**Pull request by kabili207**
This adds the location of the authenticated account to the list of people returned. Part of the data was already present in the response, it's just missing the name and photo fields. To save on extra calls, this change just sets those all to "self"

### CODE REVIEW

**Comment by costastf** [comment]
Hi! Thanks a lot for this merge request but it does not work for me. 

```ERROR:locationsharinglib.Service:Caught exception
Traceback (most recent call last):
  File "/home/costas/PycharmProjects/locationsharinglib/locationsharinglib/locationsharinglib.py", line 280, in _get_data
    output[9][1],
TypeError: 'NoneType' object has no attribute '__getitem__'
``` 
I have vigorously gone through the data returned and there is definitely no other pair of coordinates apart from the ones shared. What I imagine is that you are using this on your primary account and you have some setting enabled that make it reveal more data. Currently this has been designed to work for a "secondary" account with not settings whatsoever, apart for locations shared to it by other accounts. If you can figure out which setting makes you account reveal the extra data and we can document it and manage it in the code I would love to merge this, but currently it will just break anyone using this as it was designed to be used. I hope this makes sense and again thanks!

**Comment by costastf** [comment]
So, what we could do is handle that exception and add the current account only on success, so it works for whoever it works and does not break anyone else. It would still be nice to figure out what setting enables the extra data in case people want to do that. And may I suggest instead of using 'self' just add the email (self.email) so it is a bit more readable and explicit? if you want to make the changes I would gladly accept the MR.

ps: I added a debug of the data returned so it is easier to extend. you should probably pull if you want to work on it ;)

**Comment by kabili207** [comment]
You are correct, I was using my primary account to test this. When I tested this with another account, I got the same error you did. I had to add the secondary account to my phone and share it's location before I could get the location details of the account itself.

I also had to add an extra check because no one was sharing details with this account either.

**Comment by costastf** [comment]
I think that the setup of this is a little complex to be useful don't you?

**Comment by kabili207** [comment]
For people using a secondary account specifically for tracking, I would agree. However, this enables the use of a primary account. I would argue that it simplifies things for most users as are no longer required to set up a second account and then convince anyone else they wish to track to add said account, something which many people (such as my spouse) would see as being unnecessarily pointless.

File: locationsharinglib/locationsharinglib.py [hunk 1]
```diff
@@ -274,7 +274,26 @@ def _get_data(self):
         self._logger.debug(response.text)
         try:
             output = json.loads(response.text.split("'", 1)[1])
-            people = [Person(info) for info in output[0]]
+            people = []
+            if output[0] is not None:
```
Comment (costastf) [inline locationsharinglib/locationsharinglib.py]: Output can never be none, so checking for it does not make sense.

File: locationsharinglib/locationsharinglib.py [hunk 2]
```diff
@@ -274,7 +274,26 @@ def _get_data(self):
...
+                people.extend([Person(info) for info in output[0]])
+
+            if output[9] is not None and output[9][1] is not None:
```
Comment (costastf) [inline locationsharinglib/locationsharinglib.py]: Same here.

**Comment by costastf** [comment]
Fair enough, perfectly valid. I think that your previous mr wrapped in a try and except for that exception is much better than the current one.

File: locationsharinglib/locationsharinglib.py (hunk 1 above)
Comment (kabili207) [reply locationsharinglib/locationsharinglib.py]: Output can't be none, but the element at index 0 _can_ be none if no one is sharing their location with the account.

Comment (costastf) [reply locationsharinglib/locationsharinglib.py]: Good point.I will check a non shared account tomorrow and get back to you. Thanks!

File: locationsharinglib/locationsharinglib.py (hunk 2 above)
Comment (kabili207) [reply locationsharinglib/locationsharinglib.py]: Same as my previous comment with regards to output[9]. But I agree that the check for output[9][1] is probably unnecessary.

**Review by costastf — CHANGES_REQUESTED** [review]
All of these checks are very un-pythonic. Please have a look at https://docs.quantifiedcode.com/python-anti-patterns/readability/asking_for_permission_instead_of_forgiveness_when_working_with_files.html. If output[0] is zero it is caught on the except and the method would return an empty tuple. So basically we have three situations. 
One is that nothing is shared and nothing is returned since the exception is handled. 
Two is that third party accounts are shared and they are returned (current working implementation)
Three is that the main account is also sharing and that requires access to output[9][1]. If that is None we will have an exception that would make the code to fail to return even the ones that are shared. Checking for everything is very verbose. A nice and elegant way as I said before is to revert to your original MR which looks fine and just wrap the access to output[9][1] and the append to the persons list in a try except for TypeError so if there is no sharing of the main account just the others are returned. I hope this makes sense. :) Also please add your name or handle to the __credits__ meta and to contributors under AUTHORS.rst .

**Comment by kabili207** [comment]
Ah, this makes sense now. I'll admit, I spend most of my time with C-style languages and was not aware of this convention in Python.

There is a fourth situation that will need to be covered as well: the main account is sharing its location but no one it sharing their location with it. It's probably going to be a pretty rare scenario, but it is possible and I feel it should be covered. Perhaps if the user only wants to track themself. It's actually the current state of my test account.

**Comment by costastf** [comment]
Sure, if it is a valid state it should be covered. Not really important how rare it is, actually if we manage to cover all corner cases, the better.

**Comment by kabili207** [comment]
I believe I found a solution that's much cleaner. I've created separate methods to retrieve the locations shared to the account and the location of the account itself, and modified `get_all_people()` to combine the results of these two methods instead.

File: locationsharinglib/locationsharinglib.py [hunk 3]
```diff
@@ -272,26 +272,53 @@ def _get_data(self):
...
+        except (IndexError, TypeError):
+            self._logger.exception('Response: %s', output)
+            return None
+        return person
+
     def get_all_people(self):
-        """Retrieves all people that share their location with this account"""
-        return self._get_data()
+        """Retrieves all people sharing their location"""
+        return filter(None, self.get_shared_people() + [self.get_authenticated_person()])
```
Comment (costastf) [inline locationsharinglib/locationsharinglib.py:312]: Filter is not required and thus misleading, or am I missing something? You would just want to return the two lists as one so effectively the addition is enough.

Comment (kabili207) [reply locationsharinglib/locationsharinglib.py:312]: `get_authenticated_person()` can return None, and the filter seemed like a more elegant way of stripping it out.

Comment (costastf) [reply locationsharinglib/locationsharinglib.py:312]: oh, right, sorry missed that on the small screen of my mobile. Looks great!

**Review by costastf — APPROVED** [review]
